src/object/srv_obj.c

```diff
@@ -2743,6 +2743,17 @@ ds_obj_tgt_punch_handler(crt_rpc_t *rpc)
 	/* local RPC handler */
 	rc = obj_local_punch(opi, opc_get(rpc->cr_opc), &ioc, &dth);
 	if (rc != 0) {
+		if (rc == -DER_NONEXIST) {
```

Comment:
Hmm, the only reason you would get -DER_NONEXIST on a punch is if it's conditional, no?   One must not ignore this rc altogether.  You can safely ignore it only if there is at least one shard that gets 0 as a return code.  Otherwise, the user will need -DER_NONEXIST returned.

Reply:
hmm, the reason I ignore this is because  ec object partial update will only do updates on part of shards.  So let's say the object (4+2).  if an partial update will do update on 
1st data cell shard (0 shard) and parity shard (4,5 shards), then do a object punch, it will not be able to find the object in shard 1, 2, 3.

Reply:
I think the right thing to do here is to mask conditional flags for updates to non-parity targets and let the parity vos targets return -DER_NONEXIST/-DER_EXIST and pass those back to the user.  That way, the conditional matches the global state for the object/key.

Reply:
I discussed this with @wangdi1 and it doesn't work.   We settled on this:

For conditional punch and update, the EC operation returns -DER_NONEXIST only iff every target returns -DER_NONEXIST.  Otherwise, return 0 (or other error, if applicable).
For conditional insert, if we relax the requirements so that atomicity isn't guaranteed across shards, we can simply return -DER_EXIST if it happens and user is on their own if they send updates where extents cross shards.   Anything else, is very difficult to implement.

Reply:
argh!  I don't think that works either for conditional update/punch.  Consider this example...shard 3 has data but shard 2 doesn't.   Shard 3 will be updated but shard 2 will not.